### PR TITLE
Keras-like API Recurrent layers

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

Add Recurrent layers (SimpleRNN, LSTM, GRU, Highway) for new API.
KerasUtils redundant for several PRs, will rebase afterwards.

## How was this patch tested?

Jenkins.

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Recurrent.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+abstract class Recurrent[T: ClassTag](
+   val outputDim: Int,
+   val returnSequences: Boolean = false,
+   val goBackwards: Boolean = false,
+   var inputShape: Shape = null)(implicit ev: TensorNumeric[T])
+  extends KerasLayer[Tensor[T], Tensor[T], T](KerasLayer.addBatch(inputShape)) {
+
+  override def computeOutputShape(inputShape: Shape): Shape = {
+    val input = inputShape.toSingle().toArray
```
Comment (hkvision) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Recurrent.scala]: add require shape size

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, TensorModule}
+import com.intel.analytics.bigdl.optim.Regularizer
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Shape
+
+import scala.reflect.ClassTag
+
+class GRU[T: ClassTag](
+   outputDim: Int,
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala]: any specific reason for not adding var or val here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+   returnSequences: Boolean = false,
+   goBackwards: Boolean = false,
+   var wRegularizer: Regularizer[T] = null,
+   var uRegularizer: Regularizer[T] = null,
+   var bRegularizer: Regularizer[T] = null,
+   inputShape: Shape = null)(implicit ev: TensorNumeric[T])
+  extends Recurrent[T](outputDim, returnSequences, goBackwards, inputShape) {
+
+  override def doBuild(inputShape: Shape): AbstractModule[Tensor[T], Tensor[T], T] = {
+    val input = inputShape.toSingle().toArray
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala]: add shape validation here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Highway.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+class Highway[T: ClassTag](val activation: TensorModule[T] = null,
+                           var wRegularizer: Regularizer[T] = null,
+                           var bRegularizer: Regularizer[T] = null,
+                           val bias: Boolean = true,
+                           var inputShape: Shape = null)(implicit ev: TensorNumeric[T])
+  extends KerasLayer[Tensor[T], Tensor[T], T](KerasLayer.addBatch(inputShape))
+    with IdentityOutputShape {
+
+  override def doBuild(inputShape: Shape): AbstractModule[Tensor[T], Tensor[T], T] = {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Highway.scala]: validation

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala (hunk 2 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala]: Inherits the `Recurrent` base class. In the base class, there is val

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/LSTM.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+   val innerActivation: TensorModule[T] = null,
+   returnSequences: Boolean = false,
+   goBackwards: Boolean = false,
+   var wRegularizer: Regularizer[T] = null,
+   var uRegularizer: Regularizer[T] = null,
+   var bRegularizer: Regularizer[T] = null,
+   inputShape: Shape = null)(implicit ev: TensorNumeric[T])
+  extends Recurrent[T](outputDim, returnSequences, goBackwards, inputShape) {
+
+  override def doBuild(inputShape: Shape): AbstractModule[Tensor[T], Tensor[T], T] = {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/LSTM.scala]: validation

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasUtils.scala [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  private[keras] def getActivation[T : ClassTag] (activation: String)
+    (implicit ev: TensorNumeric[T]): TensorModule[T] = {
+    if (activation == null) null
+    else {
+      activation.toLowerCase() match {
+        case "tanh" => Tanh[T]()
+        case "sigmoid" => Sigmoid[T]()
+        case "relu" => ReLU[T]()
+        // TODO: softmax 3D input is different from Keras
+        case "softmax" => SoftMax[T]()
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasUtils.scala]: rebase

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala (hunk 3 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/GRU.scala]: Validated in `Recurrent`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/LSTM.scala (hunk 5 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/LSTM.scala]: Validated in `Recurrent`

**Comment by hkvision** [comment]
Have met all the review comments and jenkins passed http://172.168.2.101:8080/view/PR-V(new)/job/BigDL-PR-Validation-v3/113/

**Pull request merged**